framework/src/play/PropertiesConfLoader.java

```diff
@@ -32,14 +30,14 @@ public Properties readOneConfigurationFile(String playId, String filename) {
     return readOneConfigurationFile(filename, playId, null, new HashSet<>());
   }
 
-  private Properties readOneConfigurationFile(String filename, String playId, String inheritedId, Set<VirtualFile> confs) {
-    VirtualFile conf = VirtualFile.open(Play.applicationPath + "/conf/" + filename);
-    if (confs.contains(conf)) {
+  private Properties readOneConfigurationFile(String filename, String playId, String inheritedId, Set<String> confs) {
+    ClasspathResource conf = ClasspathResource.file(filename);
```

Comment:
Not adding the `/conf/` prefix to `filename` causes NPE when the `application.conf` is still at the path `/conf/application.conf`:
```
$ find build/ -name application.conf
build/resources/main/conf/application.conf
```

```
> Task :run FAILED
Exception in thread "main" java.lang.NullPointerException: File not found in classpath: application.conf
        at java.base/java.util.Objects.requireNonNull(Objects.java:347)
        at play.ClasspathResource.<init>(ClasspathResource.java:42)
        at play.ClasspathResource.file(ClasspathResource.java:26)
        at play.PropertiesConfLoader.readOneConfigurationFile(PropertiesConfLoader.java:34)
        at play.PropertiesConfLoader.readOneConfigurationFile(PropertiesConfLoader.java:30)
        at play.PropertiesConfLoader.readConfiguration(PropertiesConfLoader.java:26)
        at play.Play.readConfiguration(Play.java:231)
        at play.Play.init(Play.java:139)
        at play.server.Starter.start(Starter.java:19)
...
```

I "define" `resources` with the snippet below:
```groovy
processResources {
  from('conf') {into 'conf'}
  from('app') {include 'play.plugins'}
  from('app/views') {into 'app/views'}
  from('public') {into 'public'}
}
```

May I ask to keep that `conf/` prefix too?

Reply:
@xabolcs Sorry, maybe I don't understand, but I think there is no problems here. 
1. Thanks to the line `from('conf') {into 'conf'}` in your build script, all files in `conf` folder will be automatically in classpath. 
2. It means that the line `ClasspathResource.file("application.conf")` WILL find the file.

Reply:
Let me retest later, but for my first try the framework wasn't able to found my `application.conf` at `build/resources/main/conf/application.conf` path.
If leave out `into 'conf'`, to put it at `build/resources/main/application.conf`, then the framework starts successfully.

But let me rebase my app (and it's lib) to your recent changes first and test again.

Reply:
> 1. Thanks to the line `from('conf') {into 'conf'}` in your build script, all files in `conf` folder will be automatically in classpath.

True. I'd like to keep the structure as close as possible to Play! 1. :sweat_smile: 


> It means that the line `ClasspathResource.file("application.conf")` WILL find the file.

Does it search recursively? :wink: 
I think it doesn't.

Reply:
I don't understand what means "recursively" in this context.
But yes, when Java searches a file in classpath, it looks into ALL classpath elements (src/main/java/, src/main/resources, src/test/java etc.)

Reply:
> I don't understand what means "recursively" in this context.

I meant it as `ClasspathResource.file("application.conf");` would find `conf/application.conf`. And I think it wouldn't!

For example, with the following snippet the app starts:
```groovy
processResources {
  from('conf') {
    //into 'conf'
  }
}
```
Sure, as `application.conf` is at "root directory":
```bash
$ find build/ -name application.conf
build/resources/main/application.conf
```

---

But with the following snippet the app doesn't start:
```groovy
processResources {
  from('conf') {
    into 'conf'
  }
}
```
"File not found in classpath: application.conf"
```java
Exception in thread "main" java.lang.NullPointerException: File not found in classpath: application.conf
        at java.base/java.util.Objects.requireNonNull(Objects.java:347)
        at play.ClasspathResource.<init>(ClasspathResource.java:42)
        at play.ClasspathResource.file(ClasspathResource.java:26)
        at play.PropertiesConfLoader.readOneConfigurationFile(PropertiesConfLoader.java:34)
        at play.PropertiesConfLoader.readOneConfigurationFile(PropertiesConfLoader.java:30)
        at play.PropertiesConfLoader.readConfiguration(PropertiesConfLoader.java:26)
        at play.Play.readConfiguration(Play.java:230)
        at play.Play.init(Play.java:138)
        at play.server.Starter.start(Starter.java:19)
        at com.openseedbox.Application.main(Application.java:8)
```

And the copied `application.conf` is in "`/conf/` directory" - as requested in `build.gradle`:
```bash
$ find build/ -name application.conf
build/resources/main/conf/application.conf
```

That's what I meant for "recursively": will `ClasspathResource.file()` found all `application.conf` even when it's in `conf/` "directory"?

Reply:
Now I understand. Now, search in classpath does NOT look into subfolders. 
But why it should? Why not just copy files from `conf` to the root of classpath?

```groovy
sourceSets {
  main {
    resources {srcDir "app"}
    resources {srcDir 'conf'}
...
```

If you want to keep the original phylosophu of Play! framework, then I think they also copied `conf` files to the root of classpath. At least I found these two pieces in the original Play! code:

1. in `utils.py`:
```python
 copy_directory(os.path.join(app.path, 'conf'), os.path.join(war_path, 'WEB-INF/classes'))
```

2. in `Play.java`:
```java
 javaPath = new CopyOnWriteArrayList<VirtualFile>();
 javaPath.add(appRoot.child("app"));
 javaPath.add(appRoot.child("conf"));
```

Reply:
> Now, search in classpath does NOT look into subfolders.
> But why it should? 

Sure it should not.
But you already changing `VirtualFile conf = VirtualFile.open(Play.applicationPath + "/conf/" + filename);` into `ClasspathResource conf = ClasspathResource.file(filename);` within class `PropertiesConfLoader` in method `readOneConfigurationFile`.

At least it would be nice to use that `conf/` prefix as a fallback.

Reply:
Sorry, I still don't understand WHY it would be nice.
These two commands will find the same file `conf/application.conf`:
1. `VirtualFile.open(Play.applicationPath + "conf/application.conf")`
2. `ClasspathResource.file("application.conf");`

(assuming that `conf` folder is marked as a resource folder).
What is the problem here? What problem are you trying to solve?

Reply:
> Sorry, I still don't understand WHY it would be nice.

In short: it's all about #198

(And let me explain later why I opened it in the first place.)

Reply:
Yes, I am aware of [#198](https://github.com/replay-framework/replay/issues/198). Current PR #291 targets exactly this problem. 

But I still don't understand. You just need to remove `into 'conf'` from your build script.
Instead of this:
```
processResources {
  from('conf') {into 'conf'}
```

you need to just use `conf`:
```
sourceSets {
  main {
    resources { srcDir 'conf' }
  ...
```

Then you will not have any problems with `conf` folder.

Reply:
`resources` doesn't allow multiple `srcDir`s currently - see [How Do I Add Multiple Directories And Filter Patterns To A SourceSet?](https://stackoverflow.com/questions/63626566/how-do-i-add-multiple-directories-and-filter-patterns-to-a-sourceset)

Let me stick to `processResources` for now.

In the mean time I started to document my original issues ... [why I opened #198](https://github.com/replay-framework/replay/issues/198#issuecomment-1876230714). Let me finish that and decide if it's valid at all.

Reply:
> But I still don't understand. You just need to remove into 'conf' from your build script.

You are right, I don't have any real reason for you. It just looks better to me if a Play! like framework supports `conf/` for configuration files.

Reply:
@xabolcs Thanks. Yes, but even now I still don't pretty understand... Replay does support `conf` folder for configuration files...